bftengine/src/bftengine/Crypto.cpp

```diff
@@ -58,7 +58,18 @@ namespace impl {
 //#define RSA_STANDARD PKCS1v15
 //#define RSA_STANDARD OAEP<SHA>
 
-static RandomPool sGlobalRandGen;  // not thread-safe !!
+// Wrap RandomPool in a struct which initialises it. The wrapper is used as static thread local variable so that each
+// thread has got its own instance. The reason for this complication is thread safety. This RandomPool is used by the
+// SigManager which is a singleton used by many threads concurrently.
+struct RandomPoolWrapper : RandomPool {
+  RandomPoolWrapper() {
+    static std::atomic<uint32_t> cnt = 0;
```

Comment:
Nit: Maybe it is worth describing what `cnt` is used for and why.

Reply:
Actually, what worries me is the following (especially the last sentence):

_[basic.start.dynamic] It is implementation-defined whether the dynamic initialization of a non-local non-inline variable with thread storage duration is sequenced before the first statement of the initial function of a thread or is deferred. If it is deferred, the initialization associated with the entity for thread t is sequenced before the first non-initialization odr-use by t of any non-inline variable with thread storage duration defined in the same translation unit as the variable to be initialized. It is implementation-defined in which threads and at which points in the program such deferred dynamic initialization occurs._

I might be misunderstanding above, but I guess `cnt` will not work as intended in cases where two or more random number generators are initialised from the same thread. Maybe we can use another random seed? You were mentioning /dev/urandom or something like that.